sql/src/main/java/io/crate/execution/engine/ExecutionPhasesTask.java

```diff
@@ -197,7 +196,9 @@ private void setupContext(Map<String, Collection<NodeOperation>> operationByServ
         List<Tuple<ExecutionPhase, RowConsumer>> handlerPhaseAndReceiver = createHandlerPhaseAndReceivers(
             handlerPhases, handlerConsumers, initializationTracker);
 
-        JobExecutionContext.Builder builder = jobContextService.newBuilder(jobId, localNodeId, operationByServer.keySet());
+        JobExecutionContext.Builder builder = jobContextService.newBuilder(jobId, localNodeId)
+            .participatingNodes(operationByServer.keySet())
```

Comment:
any reason to remove the participatingNodes from the default constructor? This was intentional to make sure that all *required* arguments are passed in.

Reply:
`participatingNodes` is only set at one place (here), everywhere else, it's an empty collection, that's why I remove it from the constructor.
In general I think a Builder class should have as little required arguments as possible and everything else with common defaults, otherwise there is no need for a builder.
However, not a strong opinion in this particular case.

Reply:
> In general I think a Builder class should have as little required arguments as possible and everything else with common defaults, otherwise there is no need for a builder.

I disagree with that. It's a common use case, yes, but not the only one.
In this case the builder is used so that we've an object to pass around to which we can add sub-contexts to during the building phase. We wanted to make it impossible to add further contexts to the `JobExecutionContext` once built.

I tend to favour making the decision to pass a null/empty list at the call-site explicitly to make it clear that it was a conscious decision rather than a "I didn't see the other overload or setter".
Sure it's less convenient while writing code, but I think it tends to help while reading code. Especially tracking usages is much easier with fewer overloads